# Why shouldn't do like this??
${X}{w} = {y}$

to find ${w}$ of linear Why don't do like this?

${X^{-1}X}{w} = X^{-1}{y}$

but do like this

$({X} ^{T} {X})^{-1} {(X^{T} X)}{w} = ({X} ^{T} {X})^{-1} {X} ^ {T}{y}$

In [43]:
import numpy as np

In [44]:
X = [
    [148, 24, 1385],
     [132, 24, 2031],
     [453, 11, 86],
     [158, 24, 185],
     [172, 25,201],
     [413, 11, 86],
     [38, 54, 185],
     [142, 25, 431],
     [453, 31, 86]
]

X = np.array(X)
y = [10000, 20000, 15000, 25000, 10000, 20000, 15000, 25000, 12000]

AS ${X}{w} = {y}$ 

I inverse ${X}$  to eliminate ${X}$ in order to find ${w}$

${X}^{-1}$

but look at the error

In [45]:
try :
    np.linalg.inv(X)
except Exception as e: 
    print(e)

Last 2 dimensions of the array must be square


Yes!! I just knew that to inverse Matrix it must be sqaure.

So the solution is

add ${X^{T}}$ to ${X}$ it will be sqaure.

In [46]:
XTX = X.T.dot(X)
XTX

array([[ 696471,   43983,  718540],
       [  43983,    7097,  116772],
       [ 718540,  116772, 6359986]])

My equation will be like this

${(X^{T} X)}{w} = {X} ^ {T}{y}$

So I can eliminate ${(X^{T} X)}$ because ${(X^{T} X)}$ is sqaure now

$({X} ^{T} {X})^{-1} {(X^{T} X)}{w} = ({X} ^{T} {X})^{-1} {X} ^ {T}{y}$

$I {w} = ({X} ^{T} {X})^{-1} {X} ^ {T}{y}$

so the final ${W}$ is

In [47]:
np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)

array([ 26.20010989, 304.94155925,   3.81487275])

But actually real equation will be like this

${X}{w} + {w0} = {y}$

I will merge ${w}$ with ${w0}$

So I need to plus ${X}$  with an  ${ones}$ that have row equal to ${X}$

${(ones + X){w} = {y}}$

In [48]:
ones = np.ones(X.shape[0])
ones

array([1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [49]:
ones = np.ones(X.shape[0])
X = np.column_stack([ones, X])
list(X)

[array([1.000e+00, 1.480e+02, 2.400e+01, 1.385e+03]),
 array([1.000e+00, 1.320e+02, 2.400e+01, 2.031e+03]),
 array([  1., 453.,  11.,  86.]),
 array([  1., 158.,  24., 185.]),
 array([  1., 172.,  25., 201.]),
 array([  1., 413.,  11.,  86.]),
 array([  1.,  38.,  54., 185.]),
 array([  1., 142.,  25., 431.]),
 array([  1., 453.,  31.,  86.])]

${(ones + X){w} = {y}}$

${(ones + X)}$ can be considered as 1 Matrix

${(X){w} = {y}}$

In [50]:
XTX = X.T.dot(X)
XTX

w_full = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)

In [51]:
w0 = w_full[0]
w = w_full[1:]

print("w0 = %s" % w0)
print("w = %s" % w)

w0 = 30338.235198357772
w = [ -23.24508418 -264.48613455   -2.44929629]


So, from this equation ${X}{w} + {w0} = {y}$

The final equation after training will be like this

${X}{w} + {w0} = {y}$

${(-23.2)x_{i1}} + {(-264.5)x_{i2}} + {(-2.45)x_{i3}} + {30338.2} = y_{i}$